### PR TITLE
WIP: adds monad laws and a tool to check them

### LINKED ISSUE #533: Monad laws as values
We need to add self-checkable monad laws represented as types.

Here are some context: 
- https://github.com/zio/zio/blob/f36ba93af21f4cdcd880a39bde7911f127ff1fc6/test/shared/src/main/scala/zio/test/laws/package.scala
- https://github.com/zio/zio/blob/f963adee37635f021a741d9d5ae4ca2ed4890b10/test-tests/shared/src/test/scala/zio/test/laws/LawsSpec.scala

Related: https://github.com/dry-python/returns/issues/457

Some very early prototype:

```python
class SymmetryLaw(Law, Generic[_E]):
    def statement(self, a: _E, b: _E) -> bool:
        return a == b is b == a

    def run(self) -> NoReturn:
        # TODO: hypothesis stuff
```

### LINKED ISSUE #541: Add `SupportsEquality` interface
Not all containers support real equality check. Sometimes we need to know what exactly can be compared.
For example:
- `IO.from_value(1) == IO.from_value(1)`
- `Future.from_value(1) != Future.from_value(1)`

We can add `equals` alias to `__eq__` for types that do support real equality.
We also need to support `Equality` laws. See #533

### LINKED ISSUE #542: Separate `IOLikeN` and `IOBasedN`
`IO` can be compared with `SupportsEqual` (#541), but `Future` or `LazyIO` (#525) cannot be.
We need to split current `IOBasedN` interface into two: 
- `IOLikeN` stays as-is
- `IOBasedN` adds `SupportsEqual` interface

### LINKED ISSUE #563: Rename `_generated` to `_internal`
This name is better.

### LINKED ISSUE #564: Future objects are not re-awaitable
Currently, this code is not possible:

```python
import asyncio
from returns.future import Future

async def main():
    f = Future.from_value(1)
    assert await f.map(abs) == await f.map(int)

asyncio.run(main())
```

It raises this exception:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/sobolev/.pyenv/versions/3.8.0/lib/python3.8/asyncio/runners.py", line 43, in run
    return loop.run_until_complete(main)
  File "/Users/sobolev/.pyenv/versions/3.8.0/lib/python3.8/asyncio/base_events.py", line 608, in run_until_complete
    return future.result()
  File "<stdin>", line 3, in main
  File "/Users/sobolev/Documents/github/returns/returns/future.py", line 169, in awaitable
    return IO(await self._inner_value)
  File "/Users/sobolev/Documents/github/returns/returns/_internal/futures/_future.py", line 18, in async_map
    return function(await inner_value)
RuntimeError: cannot reuse already awaited coroutine
```

### PR DESCRIPTION
**Pull request by sobolevn**
Tada! 🎉 

TODOs for later:
- [x] add `hypothesis` to dev dependencies
- [ ] add docs for `contrib.hypothesis`
- [x] test all possible arguments to `check_all_laws`
- [x] add laws for applicative and bindable
- [x] add type tests for `SupportsEquality`
- [x] add tests for `.equals()`
- [x] add `equals` to all containers that can be compared 
- [x] add type tests for `check_all_laws`
- [x] test `.laws()` method resolution for different types
- [x] fix `typesafety/test_interfaces/test_iterable/test_inheritance.yml`
- [x] add `IOBasedN` tests

New laws to implement:
- [ ] `AltableN` must have the same laws as `MappableN`
- [ ] `.from_failure` has a law: https://wiki.haskell.org/Typeclassopedia#MonadFail
- [ ] `AsyncFutureN` has `idempotence` law
- [ ] `ReaderBasedN` must be pure 

Closes #533 
Closes #564 
Closes #563
Closes #541
Closes #542

### CODE REVIEW

**Comment by sobolevn** [comment]
Ok, it is better to switch to https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.builds and create custom strategies for our internal types.

**Comment by sobolevn** [comment]
I kindly ask @Zac-HD to help me with reviewing `hypothesis` related parts in this PR.
(Sorry, this PR is huge *for reasons* 😄 )

Here they are:
- entrypoint (not working for now, see https://github.com/HypothesisWorks/hypothesis/issues/2564) https://github.com/dry-python/returns/pull/543/files#diff-0df8bc06f8bf434789b78a0a77398673R1 
- our `check_all_laws` magic function defition: https://github.com/dry-python/returns/pull/543/files#diff-4472972bc0bac0e9127dc675dbd9f29cR1
- Equality laws definition: https://github.com/dry-python/returns/pull/543/files#diff-8aa5189ba9a3be0849ada098c397d9a8R9
- Functor laws definition: https://github.com/dry-python/returns/pull/543/files#diff-e06c48cc820df2c8762959046a397024R21
- `Law` type and friends: https://github.com/dry-python/returns/pull/543/files#diff-d07e772a790669b34462d39b35073de5R1
- Actual tests: https://github.com/dry-python/returns/pull/543/files#diff-e7433e623c9758b157f27d766eb0df03R8

File: returns/contrib/hypothesis/laws.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+
+def _run_law(
+    container_type: Type[Lawful],
+    law: Law,
+) -> Callable[[st.DataObject], None]:
+    def factory(source: st.DataObject) -> None:
+        with pure_functions():
+            with container_strategies(container_type):
+                source.draw(st.builds(law.definition))
```
Comment (Zac-HD) [inline returns/contrib/hypothesis/laws.py:122]:
```suggestion
        with pure_functions(), container_strategies(container_type):
            source.draw(st.builds(law.definition))
```

See [here](https://docs.python.org/3/reference/compound_stmts.html#with), or just run `shed --refactor` :wink:

File: returns/contrib/hypothesis/laws.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    return factory
+
+
+def _create_law_test_case(
```
Comment (Zac-HD) [inline returns/contrib/hypothesis/laws.py:126]: This is totally evil, and I love it :joy: 

People debugging might not, but...

**Review by Zac-HD — COMMENTED** [review]
This is about all I have time for, but nothing stands out as problematic.  Good luck!

File: returns/interfaces/equality.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+    @staticmethod
+    def transitivity_law(
+        first: _EqualType,
+        second: _EqualType,
+        third: _EqualType,
+    ) -> None:
+        """If ``A == B`` and ``B == C`` then ``A == C``."""
+        if first.equals(second) and second.equals(third):
+            assert first.equals(third)
```
[the remark this answers is not in the document]
Comment (sobolevn) [reply returns/interfaces/equality.py:43]: This one fails, here's an example:

```
<Failure: > <Failure: 0> <Failure: >
```

The first equals to the third one, but not the second one.

File: returns/contrib/hypothesis/_entrypoint.py [hunk 4]
```diff
@@ -0,0 +1,11 @@
+from returns.result import Result
+
+
+def hypothesis_setup_hook():
```
[the remark this answers is not in the document]
Comment (sobolevn) [reply returns/contrib/hypothesis/_entrypoint.py:4]: I will wait for https://github.com/HypothesisWorks/hypothesis/pull/2567 to get merged.

File: returns/contrib/hypothesis/laws.py (hunk 1 above)
Comment (sobolevn) [reply returns/contrib/hypothesis/laws.py:122]: We actually use an opposite rule: 

> WPS316 Found context manager with too many assignments

https://wemake-python-stylegui.de/en/latest/pages/usage/violations/consistency.html#wemake_python_styleguide.violations.consistency.MultipleContextManagerAssignmentsViolation

File: returns/contrib/hypothesis/laws.py (hunk 2 above)
Comment (sobolevn) [reply returns/contrib/hypothesis/laws.py:126]: 😄

**Comment by sobolevn** [comment]
@Zac-HD Thanks a lot for your help! 👍

**Pull request merged**